The file contains initial datasets downloading, exploring and concating in one. the final dataset is saved in data/processed/ folder

In [2]:
#!pip install seaborn

In [3]:
# import packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path


In [4]:
# set up project paths 
project_root = Path("..").resolve()
raw_dir = project_root / "data" / "raw"

train_path = raw_dir / "train.csv"
test_path = raw_dir / "test.csv"

# read the data
train = pd.read_csv(train_path)
print('train.csv is loaded!')
test = pd.read_csv(test_path)
print("test.csv is loaded!")

train.csv is loaded!
test.csv is loaded!


Let's do sanity check for the datasets before concating :)

In [5]:
# train set columns and data types
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103904 entries, 0 to 103903
Data columns (total 25 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   Unnamed: 0                         103904 non-null  int64  
 1   id                                 103904 non-null  int64  
 2   Gender                             103904 non-null  object 
 3   Customer Type                      103904 non-null  object 
 4   Age                                103904 non-null  int64  
 5   Type of Travel                     103904 non-null  object 
 6   Class                              103904 non-null  object 
 7   Flight Distance                    103904 non-null  int64  
 8   Inflight wifi service              103904 non-null  int64  
 9   Departure/Arrival time convenient  103904 non-null  int64  
 10  Ease of Online booking             103904 non-null  int64  
 11  Gate location                      1039

In [6]:
# train test first 5 rows
train.head()

Unnamed: 0      id  Gender      Customer Type  Age   Type of Travel  \
0           0   70172    Male     Loyal Customer   13  Personal Travel   
1           1    5047    Male  disloyal Customer   25  Business travel   
2           2  110028  Female     Loyal Customer   26  Business travel   
3           3   24026  Female     Loyal Customer   25  Business travel   
4           4  119299    Male     Loyal Customer   61  Business travel   

      Class  Flight Distance  Inflight wifi service  \
0  Eco Plus              460                      3   
1  Business              235                      3   
2  Business             1142                      2   
3  Business              562                      2   
4  Business              214                      3   

   Departure/Arrival time convenient  ...  Inflight entertainment  \
0                                  4  ...                       5   
1                                  2  ...                       1   
2                                  2  ...                       5   
3                                  5  ...                       2   
4                                  3  ...                       3   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 4                 3                 4                4   
1                 1                 5                 3                1   
2                 4                 3                 4                4   
3                 2                 5                 3                1   
4                 3                 4                 4                3   

   Inflight service  Cleanliness  Departure Delay in Minutes  \
0                 5            5                          25   
1                 4            1                           1   
2                 4            5                           0   
3                 4            2                          11   
4                 3            3                           0   

   Arrival Delay in Minutes             satisfaction  
0                      18.0  neutral or dissatisfied  
1                       6.0  neutral or dissatisfied  
2                       0.0                satisfied  
3                       9.0  neutral or dissatisfied  
4                       0.0                satisfied  

[5 rows x 25 columns]

In [7]:
# test dataset columns and datatypes
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25976 entries, 0 to 25975
Data columns (total 25 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Unnamed: 0                         25976 non-null  int64  
 1   id                                 25976 non-null  int64  
 2   Gender                             25976 non-null  object 
 3   Customer Type                      25976 non-null  object 
 4   Age                                25976 non-null  int64  
 5   Type of Travel                     25976 non-null  object 
 6   Class                              25976 non-null  object 
 7   Flight Distance                    25976 non-null  int64  
 8   Inflight wifi service              25976 non-null  int64  
 9   Departure/Arrival time convenient  25976 non-null  int64  
 10  Ease of Online booking             25976 non-null  int64  
 11  Gate location                      25976 non-null  int

In [8]:
# test set first 5 rows
test.head()

Unnamed: 0     id  Gender      Customer Type  Age   Type of Travel  \
0           0  19556  Female     Loyal Customer   52  Business travel   
1           1  90035  Female     Loyal Customer   36  Business travel   
2           2  12360    Male  disloyal Customer   20  Business travel   
3           3  77959    Male     Loyal Customer   44  Business travel   
4           4  36875  Female     Loyal Customer   49  Business travel   

      Class  Flight Distance  Inflight wifi service  \
0       Eco              160                      5   
1  Business             2863                      1   
2       Eco              192                      2   
3  Business             3377                      0   
4       Eco             1182                      2   

   Departure/Arrival time convenient  ...  Inflight entertainment  \
0                                  4  ...                       5   
1                                  1  ...                       4   
2                                  0  ...                       2   
3                                  0  ...                       1   
4                                  3  ...                       2   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 5                 5                 5                2   
1                 4                 4                 4                3   
2                 4                 1                 3                2   
3                 1                 1                 1                3   
4                 2                 2                 2                4   

   Inflight service  Cleanliness  Departure Delay in Minutes  \
0                 5            5                          50   
1                 4            5                           0   
2                 2            2                           0   
3                 1            4                           0   
4                 2            4                           0   

   Arrival Delay in Minutes             satisfaction  
0                      44.0                satisfied  
1                       0.0                satisfied  
2                       0.0  neutral or dissatisfied  
3                       6.0                satisfied  
4                      20.0                satisfied  

[5 rows x 25 columns]

In [9]:
# concating the sets
df=pd.concat([train, test], axis=0, ignore_index=True)

In [10]:
# check
print(f'full set has rows: {df.shape[0]}')
print(f'... and columns: {df.shape[1]}')


full set has rows: 129880
... and columns: 25


In [11]:
# the column 'Unamed is redundant - duplicating rows count
df.head()

Unnamed: 0      id  Gender      Customer Type  Age   Type of Travel  \
0           0   70172    Male     Loyal Customer   13  Personal Travel   
1           1    5047    Male  disloyal Customer   25  Business travel   
2           2  110028  Female     Loyal Customer   26  Business travel   
3           3   24026  Female     Loyal Customer   25  Business travel   
4           4  119299    Male     Loyal Customer   61  Business travel   

      Class  Flight Distance  Inflight wifi service  \
0  Eco Plus              460                      3   
1  Business              235                      3   
2  Business             1142                      2   
3  Business              562                      2   
4  Business              214                      3   

   Departure/Arrival time convenient  ...  Inflight entertainment  \
0                                  4  ...                       5   
1                                  2  ...                       1   
2                                  2  ...                       5   
3                                  5  ...                       2   
4                                  3  ...                       3   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 4                 3                 4                4   
1                 1                 5                 3                1   
2                 4                 3                 4                4   
3                 2                 5                 3                1   
4                 3                 4                 4                3   

   Inflight service  Cleanliness  Departure Delay in Minutes  \
0                 5            5                          25   
1                 4            1                           1   
2                 4            5                           0   
3                 4            2                          11   
4                 3            3                           0   

   Arrival Delay in Minutes             satisfaction  
0                      18.0  neutral or dissatisfied  
1                       6.0  neutral or dissatisfied  
2                       0.0                satisfied  
3                       9.0  neutral or dissatisfied  
4                       0.0                satisfied  

[5 rows x 25 columns]

In [12]:
# changing 'Customer type' to 'Loyalty' with yes/no 

df['Loyalty Program'] = df['Customer Type'].map({
    'Loyal Customer': 'Yes',
    'disloyal Customer': 'No'
})

# check the result
df[['Customer Type', 'Loyalty Program']].sample(10)

Customer Type Loyalty Program
88498  Loyal Customer             Yes
33312  Loyal Customer             Yes
52511  Loyal Customer             Yes
31316  Loyal Customer             Yes
37682  Loyal Customer             Yes
50385  Loyal Customer             Yes
28744  Loyal Customer             Yes
15232  Loyal Customer             Yes
50234  Loyal Customer             Yes
22274  Loyal Customer             Yes

### Duplicates check

In [13]:
# checking duplicates
key_cols = [
    'Gender',
    'Age',
    'Customer Type',
    'Loyalty Program',
    'Type of Travel',
    'Class',
    'Flight Distance',
    'Departure Delay in Minutes',
    'Arrival Delay in Minutes'
]

dupl_key = int(df.duplicated(subset=key_cols).sum())
print(f'there are {dupl_key} duplicates by key colums: \n {key_cols}')

there are 4572 duplicates by key colums: 
 ['Gender', 'Age', 'Customer Type', 'Loyalty Program', 'Type of Travel', 'Class', 'Flight Distance', 'Departure Delay in Minutes', 'Arrival Delay in Minutes']


In [14]:
# full duplicates: 
dupl_full = df.duplicated().sum()
print(f'there are {dupl_full} full duplicates')

there are 0 full duplicates


In [15]:
# lets expect duplicates
dups_key = df[df.duplicated(subset=key_cols, keep=False)].sort_values(key_cols)
dups_key.head(20)


Unnamed: 0      id  Gender   Customer Type  Age   Type of Travel  \
90354        90354   50520  Female  Loyal Customer    7  Personal Travel   
119421       15517   51021  Female  Loyal Customer    7  Personal Travel   
46641        46641   75044  Female  Loyal Customer    7  Personal Travel   
57230        57230   62336  Female  Loyal Customer    7  Personal Travel   
75843        75843  102412  Female  Loyal Customer    7  Personal Travel   
14271        14271  127767  Female  Loyal Customer    7  Personal Travel   
109741        5837  124323  Female  Loyal Customer    7  Personal Travel   
20241        20241   38885  Female  Loyal Customer    7  Personal Travel   
108204        4300   30104  Female  Loyal Customer    7  Personal Travel   
8466          8466   22872  Female  Loyal Customer    7  Personal Travel   
28318        28318   22874  Female  Loyal Customer    7  Personal Travel   
61786        61786   22851  Female  Loyal Customer    7  Personal Travel   
31603        31603  116051  Female  Loyal Customer    7  Personal Travel   
91590        91590  129784  Female  Loyal Customer    7  Personal Travel   
83975        83975  127174  Female  Loyal Customer    7  Personal Travel   
126606       22702  128480  Female  Loyal Customer    7  Personal Travel   
103469      103469  107828  Female  Loyal Customer    7  Personal Travel   
103591      103591  107841  Female  Loyal Customer    7  Personal Travel   
38062        38062  105392  Female  Loyal Customer    7  Personal Travel   
93060        93060   36439  Female  Loyal Customer    7  Personal Travel   

       Class  Flight Distance  Inflight wifi service  \
90354    Eco               77                      4   
119421   Eco               77                      1   
46641    Eco              236                      1   
57230    Eco              236                      2   
75843    Eco              236                      1   
14271    Eco              255                      2   
109741   Eco              255                      2   
20241    Eco              261                      2   
108204   Eco              261                      3   
8466     Eco              302                      3   
28318    Eco              302                      1   
61786    Eco              302                      5   
31603    Eco              337                      1   
91590    Eco              337                      5   
83975    Eco              338                      4   
126606   Eco              338                      1   
103469   Eco              349                      3   
103591   Eco              349                      4   
38062    Eco              369                      2   
93060    Eco              369                      2   

        Departure/Arrival time convenient  ...  On-board service  \
90354                                   5  ...                 5   
119421                                  3  ...                 4   
46641                                   5  ...                 3   
57230                                   5  ...                 1   
75843                                   5  ...                 2   
14271                                   4  ...                 1   
109741                                  4  ...                 5   
20241                                   1  ...                 4   
108204                                  1  ...                 1   
8466                                    1  ...                 3   
28318                                   5  ...                 4   
61786                                   2  ...                 2   
31603                                   4  ...                 5   
91590                                   4  ...                 5   
83975                                   5  ...                 3   
126606                                  5  ...                 5   
103469                                  4  ...                 5   
103591    

In [16]:
dups_key.describe()

Unnamed: 0             id          Age  Flight Distance  \
count    8661.000000    8661.000000  8661.000000      8661.000000   
mean    44585.178848   65016.116730    38.675673       888.248932   
std     31089.189480   36836.768966    14.482441       818.834397   
min         2.000000      19.000000     7.000000        67.000000   
25%     16827.000000   33160.000000    26.000000       329.000000   
50%     39775.000000   64938.000000    39.000000       581.000000   
75%     71872.000000   91904.000000    50.000000      1121.000000   
max    103848.000000  129863.000000    79.000000      4983.000000   

       Inflight wifi service  Departure/Arrival time convenient  \
count            8661.000000                        8661.000000   
mean                2.759612                           3.024939   
std                 1.356002                           1.582402   
min                 0.000000                           0.000000   
25%                 2.000000                           2.000000   
50%                 3.000000                           3.000000   
75%                 4.000000                           4.000000   
max                 5.000000                           5.000000   

       Ease of Online booking  Gate location  Food and drink  Online boarding  \
count             8661.000000    8661.000000     8661.000000      8661.000000   
mean                 2.737675       2.949659        3.215333         3.258053   
std                  1.410452       1.264366        1.330552         1.385768   
min                  0.000000       0.000000        1.000000         0.000000   
25%                  2.000000       2.000000        2.000000         2.000000   
50%                  3.000000       3.000000        3.000000         4.000000   
75%                  4.000000       4.000000        4.000000         4.000000   
max                  5.000000       5.000000        5.000000         5.000000   

       Seat comfort  Inflight entertainment  On-board service  \
count   8661.000000             8661.000000       8661.000000   
mean       3.445791                3.392911          3.491745   
std        1.340427                1.335167          1.253755   
min        1.000000                0.000000          0.000000   
25%        2.000000                2.000000          3.000000   
50%        4.000000                4.000000          4.000000   
75%        5.000000                4.000000          4.000000   
max        5.000000                5.000000          5.000000   

       Leg room service  Baggage handling  Checkin service  Inflight service  \
count       8661.000000       8661.000000      8661.000000       8661.000000   
mean           3.399030          3.705808         3.397991          3.718970   
std            1.308303          1.173239         1.239325          1.168484   
min            0.000000          1.000000         1.000000          0.000000   
25%            2.000000          3.000000         3.000000          3.000000   
50%            4.000000          4.000000         4.000000          4.000000   
75%            4.000000          5.000000         4.000000          5.000000   
max            5.000000          5.000000         5.000000          5.000000   

       Cleanliness  Departure Delay in Minutes  Arrival Delay in Minutes  
count  8661.000000                 8661.000000               8661.000000  
mean      3.300658                    0.011546                  0.019628  
std       1.315351                    0.168832                  0.472166  
min       1.000000                    0.000000                  0.000000  
25%       2.000000                    0.000000                  0.000000  
50%       3.000000                    0.000000                  0.000000  
75%       4.000000                    0.000000                  0.000000  
max       5.000000                    6.000000                 21.000000

***Duplicate checks and repeated profiles***

To check for potential duplicates, I looked at rows that share the same combination of key characteristics: gender, age, customer type, type of travel, class, flight distance, and departure/arrival delays. This produced about 4.6K rows that are flagged as duplicates *by this key*, meaning that there are many passengers who fall into identical profile combinations.

This does **not** mean that rows are literally copied. In this dataset, most variables are discrete and low-cardinality (age in whole years, service ratings from 0–5, delays in minutes, only a few travel classes and customer types). With thousands of observations, it is expected that many passengers share exactly the same profile across these fields. In other words, the duplicate check is mostly detecting **repeated passenger types**, not data entry errors.

I also inspected example rows from these groups and found that other fields (such as detailed service ratings or flight distances in the wider dataset) vary across records, and there is no evidence of inconsistent labels for clearly identical entries. Based on this, I treat these cases as normal repeated profiles rather than true duplicate records, and I do **not** drop them at this stage.


### Unique values for each column

In [17]:
df.nunique()

Unnamed: 0                           103904
id                                   129880
Gender                                    2
Customer Type                             2
Age                                      75
Type of Travel                            2
Class                                     3
Flight Distance                        3821
Inflight wifi service                     6
Departure/Arrival time convenient         6
Ease of Online booking                    6
Gate location                             6
Food and drink                            6
Online boarding                           6
Seat comfort                              6
Inflight entertainment                    6
On-board service                          6
Leg room service                          6
Baggage handling                          5
Checkin service                           6
Inflight service                          6
Cleanliness                               6
Departure Delay in Minutes      

In [18]:
# check ratings for satisfuction - for better understanding
df['Cleanliness'].unique()

array([5, 1, 2, 3, 4, 0])

In [19]:
# 'Baggage handling' - 5 - why is that?

df['Baggage handling'].unique()

array([4, 3, 5, 1, 2])

The **Baggage handling** column is the only service-related variable recorded exclusively on a 1–5 scale, with no zero values observed. All other service rating variables use a 0–5 scale. This contrast suggests that a value of 0 in the other service columns likely represents a “not applicable” or “not used” case rather than a poor service rating. Based on this interpretation, zero values in service ratings are treated as non-applicable observations rather than negative evaluations.


### Check for missing values

In [20]:
df.isna().sum()

Unnamed: 0                             0
id                                     0
Gender                                 0
Customer Type                          0
Age                                    0
Type of Travel                         0
Class                                  0
Flight Distance                        0
Inflight wifi service                  0
Departure/Arrival time convenient      0
Ease of Online booking                 0
Gate location                          0
Food and drink                         0
Online boarding                        0
Seat comfort                           0
Inflight entertainment                 0
On-board service                       0
Leg room service                       0
Baggage handling                       0
Checkin service                        0
Inflight service                       0
Cleanliness                            0
Departure Delay in Minutes             0
Arrival Delay in Minutes             393
satisfaction    

A small number of observations (393) have missing values in the **arrival delay** variable. These records were retained during exploratory analysis to understand the pattern of missingness. This missing observation will not be removed or impute before EDA. 

### Note on dataset complexity

Although the Airline Passenger Satisfaction dataset initially appears to be a straightforward survey dataset, deeper exploration reveals a richer structure. The data combines passenger demographics, travel context, operational performance (such as delays and flight distance), and multiple service quality measures with different measurement conventions. Some service features are always applicable, while others include explicit non-applicable values, and there are edge cases such as very young passengers labeled as loyal customers. Together, these characteristics indicate that the dataset supports more nuanced analysis than a simple satisfaction prediction task.

#### Drop the redundant columns and save the set

In [23]:
df_after_drop=df.drop(columns=['Unnamed: 0', 'Customer Type'])
df_after_drop.head()

id  Gender  Age   Type of Travel     Class  Flight Distance  \
0   70172    Male   13  Personal Travel  Eco Plus              460   
1    5047    Male   25  Business travel  Business              235   
2  110028  Female   26  Business travel  Business             1142   
3   24026  Female   25  Business travel  Business              562   
4  119299    Male   61  Business travel  Business              214   

   Inflight wifi service  Departure/Arrival time convenient  \
0                      3                                  4   
1                      3                                  2   
2                      2                                  2   
3                      2                                  5   
4                      3                                  3   

   Ease of Online booking  Gate location  ...  On-board service  \
0                       3              1  ...                 4   
1                       3              3  ...                 1   
2                       2              2  ...                 4   
3                       5              5  ...                 2   
4                       3              3  ...                 3   

   Leg room service  Baggage handling  Checkin service  Inflight service  \
0                 3                 4                4                 5   
1                 5                 3                1                 4   
2                 3                 4                4                 4   
3                 5                 3                1                 4   
4                 4                 4                3                 3   

   Cleanliness  Departure Delay in Minutes  Arrival Delay in Minutes  \
0            5                          25                      18.0   
1            1                           1                       6.0   
2            5                           0                       0.0   
3            2                          11                       9.0   
4            3                           0                       0.0   

              satisfaction  Loyalty Program  
0  neutral or dissatisfied              Yes  
1  neutral or dissatisfied               No  
2                satisfied              Yes  
3  neutral or dissatisfied              Yes  
4                satisfied              Yes  

[5 rows x 24 columns]

In [24]:
# save file to repo
df_after_drop.to_csv('../data/processed/airlines_full.csv', index=False)